### ISSUE TITLE
dom loader memory leak

### ISSUE THREAD

**Issue by eouna** [paquettg/php-html-parser#213]
When i running in the TSR program, the dom loader have memory leak
`  

                $sMemory = number_format((float)(memory_get_usage() / 1048576), 2);  

                dump_vars("Cur Process ID：". posix_getpid() ."Occupation Memory：" . $sMemory . 'Mb');  

                $dom = $this->dom->loadStr($response);  

                $sMemory = number_format((float)(memory_get_usage() / 1048576), 2);  

                dump_vars("Cur Process ID：". posix_getpid() ."Occupation Memory：" . $sMemory . 'Mb');  

                $siteTitle = getTitle($response);  

                !empty($this->siteName) ?: $this->siteName = mb_substr($siteTitle, (int)(strripos($siteTitle, '_')));  

                $this->dealLink($dom, $siteTitle);  

                $this->resourceDispatcher($dom, $siteTitle);  

                unset($dom);  

                $sMemory = number_format((float)(memory_get_usage() / 1048576), 2);  

                dump_vars("Cur Process ID：". posix_getpid() ."Occupation Memory：" . $sMemory . 'Mb');  

`
also, when '__destruct' i destroy the self dom object  

as you see：  

![image](https://user-images.githubusercontent.com/54256285/76492615-a2925a00-646b-11ea-8cea-42c6a196c55c.png)  

![image](https://user-images.githubusercontent.com/54256285/76492632-aa51fe80-646b-11ea-9b06-341cb69ab901.png)  

every time run in there while  increase take about 1-2 Mb memory, and then unset this entity, but it`s seem not work~  

I`am so hard╮(╯﹏╰）╭

**Comment by sergeolkhovik** [comment]
Got a similar problem but probably the reason is another. Anyway let me provide more info on this - the problem is related to incorrect HTML (3rd paragraph has no closing tag and image tag is malformed), but I see no other way to check it than try to load it. So here's the sample code:

```
<?php

# autoload etc..

$html = <<<'HTML'
<div class="ContentBoxPanel"><div class="HTMLType"><p>Test.&nbsp;</p>
<p><strong>Test:</strong> test.&nbsp;</p>
<p><img class="" src="/x/Screen%20Shot.png" width="254" height="263" />&nbsp; &nbsp; &nbsp;&nbsp;<img class="" src="data:image/png;base64,iVBVo7wmZNnbizwc/WcP</div></div>
HTML;

$now = time();
set_error_handler(function($code, $string, $file, $line) {
    print(date("H:i:s") . " Error '$string' ($code) at line $line\n");
});
register_shutdown_function(function() use ($now) {
    print(date("H:i:s") . " Done: " . (time() - $now) . " seconds\n");
});

print(date("H:i:s") . " Starting..\n");
$sMemory = number_format((float)(memory_get_usage() / 1048576), 2);  
print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n"); 
$dom = new PHPHtmlParser\Dom;
$dom->setOptions([
    // 'strict' => true, // Set a global option to enable strict html parsing.
    'cleanupInput' => false,
]);
$dom->loadStr($html);
print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n"); 

```

And here's the output:

```
# php xi.php
15:04:11 Starting..
Cur Process ID：395, Occupation Memory：1.17Mb

Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 65011720 bytes) in /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php on line 759

Call Stack:
    0.0006     360576   1. {main}() /var/www/html/xi.php:0
    0.0392    1339480   2. PHPHtmlParser\Dom->loadStr() /var/www/html/xi.php:28
    0.0430    1370720   3. PHPHtmlParser\Dom->parse() /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php:229
    0.0600    1555504   4. PHPHtmlParser\Dom->parseTag() /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php:628

15:17:48 Done: 817 seconds
```

817 seconds = 13:47, so long time..
And when running under Apache code, it even blocks other requests.
Any ideas how to handle that?
Thanks!

**Comment by eouna** [comment]
Hi body, your variable $html not close img tag,  when you close img tag, it will working
there is code:
<?php
  
# autoload etc..
require 'vendor/autoload.php';

$html = <<<'HTML'
        <div class="ContentBoxPanel">
    <div class="HTMLType"><p>Test.&nbsp;</p>
        <p><strong>Test:</strong> test.&nbsp;</p>
        <p>
            <img class="" src="/x/Screen%20Shot.png" width="254" height="263"/>&nbsp; &nbsp; &nbsp;&nbsp;
            <img class="" src="data:image/png;base64,iVBVo7wmZNnbizwc/WcP"/>
    </div>
</div>
HTML;

$now = time();
set_error_handler(function($code, $string, $file, $line) {
            print(date("H:i:s") . " Error '$string' ($code) at line $line\n");
});
register_shutdown_function(function() use ($now) {
            print(date("H:i:s") . " Done: " . (time() - $now) . " seconds\n");
});

print(date("H:i:s") . " Starting..\n");
$sMemory = number_format((float)(memory_get_usage() / 1048576), 2);
print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n");
$dom = new PHPHtmlParser\Dom;
$dom->setOptions([
            // 'strict' => true, // Set a global option to enable strict html parsing.
        'cleanupInput' => false,
    ]);
$dom->loadStr($html);
print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n");

![image](https://user-images.githubusercontent.com/54256285/82986054-b7709980-a027-11ea-9560-f24daebd2340.png)


> Got a similar problem but probably the reason is another. Anyway let me provide more info on this - the problem is related to incorrect HTML (3rd paragraph has no closing tag and image tag is malformed), but I see no other way to check it than try to load it. So here's the sample code:
> 
> ```
> <?php
> 
> # autoload etc..
> 
> $html = <<<'HTML'
> <div class="ContentBoxPanel"><div class="HTMLType"><p>Test.&nbsp;</p>
> <p><strong>Test:</strong> test.&nbsp;</p>
> <p><img class="" src="/x/Screen%20Shot.png" width="254" height="263" />&nbsp; &nbsp; &nbsp;&nbsp;<img class="" src="data:image/png;base64,iVBVo7wmZNnbizwc/WcP</div></div>
> HTML;
> 
> $now = time();
> set_error_handler(function($code, $string, $file, $line) {
>     print(date("H:i:s") . " Error '$string' ($code) at line $line\n");
> });
> register_shutdown_function(function() use ($now) {
>     print(date("H:i:s") . " Done: " . (time() - $now) . " seconds\n");
> });
> 
> print(date("H:i:s") . " Starting..\n");
> $sMemory = number_format((float)(memory_get_usage() / 1048576), 2);  
> print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n"); 
> $dom = new PHPHtmlParser\Dom;
> $dom->setOptions([
>     // 'strict' => true, // Set a global option to enable strict html parsing.
>     'cleanupInput' => false,
> ]);
> $dom->loadStr($html);
> print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n"); 
> ```
> 
> And here's the output:
> 
> ```
> # php xi.php
> 15:04:11 Starting..
> Cur Process ID：395, Occupation Memory：1.17Mb
> 
> Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 65011720 bytes) in /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php on line 759
> 
> Call Stack:
>     0.0006     360576   1. {main}() /var/www/html/xi.php:0
>     0.0392    1339480   2. PHPHtmlParser\Dom->loadStr() /var/www/html/xi.php:28
>     0.0430    1370720   3. PHPHtmlParser\Dom->parse() /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php:229
>     0.0600    1555504   4. PHPHtmlParser\Dom->parseTag() /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php:628
> 
> 15:17:48 Done: 817 seconds
> ```
> 
> 817 seconds = 13:47, so long time..
> And when running under Apache code, it even blocks other requests.
> Any ideas how to handle that?
> Thanks!

**Comment by eouna** [comment]
> Got a similar problem but probably the reason is another. Anyway let me provide more info on this - the problem is related to incorrect HTML (3rd paragraph has no closing tag and image tag is malformed), but I see no other way to check it than try to load it. So here's the sample code:
> 
> ```
> <?php
> 
> # autoload etc..
> 
> $html = <<<'HTML'
> <div class="ContentBoxPanel"><div class="HTMLType"><p>Test.&nbsp;</p>
> <p><strong>Test:</strong> test.&nbsp;</p>
> <p><img class="" src="/x/Screen%20Shot.png" width="254" height="263" />&nbsp; &nbsp; &nbsp;&nbsp;<img class="" src="data:image/png;base64,iVBVo7wmZNnbizwc/WcP</div></div>
> HTML;
> 
> $now = time();
> set_error_handler(function($code, $string, $file, $line) {
>     print(date("H:i:s") . " Error '$string' ($code) at line $line\n");
> });
> register_shutdown_function(function() use ($now) {
>     print(date("H:i:s") . " Done: " . (time() - $now) . " seconds\n");
> });
> 
> print(date("H:i:s") . " Starting..\n");
> $sMemory = number_format((float)(memory_get_usage() / 1048576), 2);  
> print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n"); 
> $dom = new PHPHtmlParser\Dom;
> $dom->setOptions([
>     // 'strict' => true, // Set a global option to enable strict html parsing.
>     'cleanupInput' => false,
> ]);
> $dom->loadStr($html);
> print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n"); 
> ```
> 
> And here's the output:
> 
> ```
> # php xi.php
> 15:04:11 Starting..
> Cur Process ID：395, Occupation Memory：1.17Mb
> 
> Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 65011720 bytes) in /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php on line 759
> 
> Call Stack:
>     0.0006     360576   1. {main}() /var/www/html/xi.php:0
>     0.0392    1339480   2. PHPHtmlParser\Dom->loadStr() /var/www/html/xi.php:28
>     0.0430    1370720   3. PHPHtmlParser\Dom->parse() /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php:229
>     0.0600    1555504   4. PHPHtmlParser\Dom->parseTag() /var/www/html/vendor/paquettg/php-html-parser/src/PHPHtmlParser/Dom.php:628
> 
> 15:17:48 Done: 817 seconds
> ```
> 
> 817 seconds = 13:47, so long time..
> And when running under Apache code, it even blocks other requests.
> Any ideas how to handle that?
> Thanks!

in your code 3rd paragraph, you close that tag it while works
`

<?php
  
# autoload etc..
require 'vendor/autoload.php';

$html = <<<'HTML'
        <div class="ContentBoxPanel">
    <div class="HTMLType"><p>Test.&nbsp;</p>
        <p><strong>Test:</strong> test.&nbsp;</p>
        <p>
            <img class="" src="/x/Screen%20Shot.png" width="254" height="263"/>&nbsp; &nbsp; &nbsp;&nbsp;
            <img class="" src="data:image/png;base64,iVBVo7wmZNnbizwc/WcP"/>
    </div>
</div>
HTML;

$now = time();
set_error_handler(function($code, $string, $file, $line) {
            print(date("H:i:s") . " Error '$string' ($code) at line $line\n");
});
register_shutdown_function(function() use ($now) {
            print(date("H:i:s") . " Done: " . (time() - $now) . " seconds\n");
});

print(date("H:i:s") . " Starting..\n");
$sMemory = number_format((float)(memory_get_usage() / 1048576), 2);
print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n");
$dom = new PHPHtmlParser\Dom;
$dom->setOptions([
            // 'strict' => true, // Set a global option to enable strict html parsing.
        'cleanupInput' => false,
    ]);
$dom->loadStr($html);
print("Cur Process ID：". posix_getpid() .", Occupation Memory：" . $sMemory . "Mb\n");

`
![image](https://user-images.githubusercontent.com/54256285/82986248-19c99a00-a028-11ea-93a9-d1bc5f054842.png)

**Comment by sergeolkhovik** [comment]
@eouna right, if I correct that then all works as expexted. But I was planning to use this library to validate HTML (user input) before saving into DB. Is that possible?

**Comment by eouna** [comment]
if your project not in resident memory environment (such as NGINX+PHP-FPM or APACHE ) and not so much request , it`s while works . In contrast your application in resident memory environment , you may consider use another package in your project. there are package may be useful ( composer package **imangazaliev/didom** ) in that situation

**Comment by eouna** [comment]
@sergeolkhovik

**Comment by eouna** [comment]
@sergeolkhovik validate HTML use input value ? if you just want to validate user input value use JS async method (ajax, fetch or axios ) delivery user value will a better choose or use HTML tag <form/> to synchronize validate input value.
Otherwise, if your project not in resident memory environment (such as NGINX+PHP-FPM or APACHE ) and not so much request , it`s will works . In contrast your application in resident memory environment , you may consider use another package in your project. there are package may be useful ( composer package imangazaliev/didom ) in that situation

**Comment by paquettg** [comment]
This issue should be resolved in 3.0.0. Please try version 3.0.0 and re-open the ticket if the issue still exists.
